eng/scripts/Generate-Dependency-Test-References.ps1

```diff
@@ -0,0 +1,94 @@
+
+.PARAMETER ProjectFileConfigName
+The config name of the project file name. Hardcoded in "\eng\pipelines\templates\jobs\ci.yml" `PreGenerationSteps` for matix generation. E.g. 'ProjectListOverrideFile' 
+
+.PARAMETER ExcludeTargetTestProjects
+The flag to indicate whether exclude target test package or not. 
+
+.PARAMETER ServiceDirectory
+The service directory which needs to exclude. E.g. "core"
+
+.EXAMPLE
+Run script with default parameters.
+
+Generate-Dependency-Test-References.ps1 -ProjectFilesOutputFolder "$(Build.ArtifactStagingDirectory)/projects" -ProjectListFilePath "eng/projects.txt" `
+-MatrixConfigsFile "eng/pipelines/templates/stages/platform-matrix.json" -ProjectFileConfigName "ProjectListOverrideFile"
+
+#>
+Param (
+    [string] $ProjectListFilePath,
+    [string] $ProjectFilesOutputFolder,
+    [int] $NumOfTestProjectsPerJob = 20,
+    [string] $MatrixConfigsFile,
+    [string] $ProjectFileConfigName,
+    [boolean] $ExcludeTargetTestProjects,
+    [string] $ServiceDirectory
+)
+
+$packages = Get-Content $ProjectListFilePath
+if ($ExcludeTargetTestProjects) {
+  $excludedPackages = @()
+  foreach ($pkg in $packages) {
+    if (!($pkg -match "sdk[\\|/]$ServiceDirectory[\\|/]")) {
+      $excludedPackages += $pkg
+    }
+  }
+  $packages = $excludedPackages
+}
+
+$numOfGroups = [Math]::Floor([decimal]($packages.Count / $NumOfTestProjectsPerJob ))
+$pkgGroup = [System.Collections.ArrayList]::new()
+$i = 0
+for ($i = 0; $i -lt $numOfGroups; $i++) { 
+  $pkgGroup.Add($packages[( $i * $NumOfTestProjectsPerJob )..(($i + 1) * $NumOfTestProjectsPerJob )])
+}
+if ($packages.Count -gt $numOfGroups * $NumOfTestProjectsPerJob) {
+  $pkgGroup.Add($packages[($numOfGroups * $NumOfTestProjectsPerJob)..$packages.Count])
+}
+
+$null = New-Item -Path $ProjectFilesOutputFolder -ItemType "directory" -Force 
+
+for ($i = 0; $i -lt $numOfGroups; $i++) {
+  $projectFilePath = "$ProjectFilesOutputFolder/Project_$i.props"
+  $null = New-Item -Path $projectFilePath -ItemType "file"
+  # Retain the structure without ProjectReference.
+  $templateXml=[xml]'<Project></Project>'  
+  $projectNode = $templateXml.SelectNodes("/Project")
+  $itemGroupNode=$templateXml.CreateNode("element", "ItemGroup", "")
+  $projectNode.AppendChild($itemGroupNode)
+  foreach($pkg in $pkgGroup[$i]) {
+    $newElem = $templateXml.CreateNode("element", "ProjectReference", "")
+    $newElemAttr = $templateXml.CreateAttribute("Include")
+    $newElemAttr.InnerText = $pkg
+    $newElem.Attributes.Append($newElemAttr)
+    
+    $itemGroupNode.AppendChild($newElem)
+    $templateXml.Save($projectFilePath)
+  }
+}
+
+$platformJson = Get-Content $MatrixConfigsFile | ConvertFrom-Json
+$projectFiles = Get-ChildItem "$ProjectFilesOutputFolder\Project_*.props"
+$platformJson.matrix | Add-Member -Name $ProjectFileConfigName -value $projectFiles.Name -MemberType NoteProperty
+$platformJson.PSObject.Properties.Remove("Include")
```

Comment:
Might be worth calling out why you are removing include. @benbp for context anything in an include group will not have a project file which means it will not really work out for chunking. I think for our scenarios here we can just remove anything not in the core matrix setup.

Reply:
I don't think you can make the call from this layer that there are no important test SKUs specified in include though?

Reply:
Pseudo-code, but I mentioned to @sima-zhu that it would probably be better to include the "include" entries in the project list sharding:

```
$platformJson = Get-Content $MatrixConfigsFile | ConvertFrom-Json
$projectFiles = Get-ChildItem "$ProjectFilesOutputFolder\Project_*.props"
foreach ($include in $platformJson.include) {
  $include | Add-Member -Name $ProjectFileConfigName -value $projectFiles.Pop().Name -MemberType NoteProperty
}
$platformJson.matrix | Add-Member -Name $ProjectFileConfigName -value $projectFiles.Name -MemberType NoteProperty
```

That said, if a config ever has an `exclude` directive, it will break things because it will remove one of the project list entries from the final matrix. I think we need to add something like "import and combine sparse" which we have resisted thus far, but it will simplify and solve this problem and some others we have seen in the past. Bad naming, but something like

```
{
  "matrix": {
    "$IMPORT": "my-matrix.json",
    "$IMPORT_COMBINE_MODE": "sparse",
    "ProjectListOverrideFile": [
      ...
    }
}
```